# 2、C++11 线程函数中的数据未定义错误

---

## 1. 传递临时变量的问题

```cpp
#include <iostream>
#include <thread>

void foo(int& x) {
    x += 1;
}

int main() {
    std::thread t(foo, 1); // 传递临时变量
    t.join();
    return 0;
}
```

### 问题分析

在上述代码中，`foo` 函数接受一个整数引用作为参数，并对其加 1。但在线程创建时，传入的 `1` 是一个临时变量，在 `std::thread` 解析参数时，该临时变量会被销毁，导致 `foo` 访问了已销毁的对象，产生未定义行为。

### 解决方案

使用 `std::ref` 传递一个持久化变量的引用：

```cpp
#include <iostream>
#include <thread>
#include <functional>

void foo(int& x) {
    x += 1;
}

int main() {
    int x = 1;
    std::thread t(foo, std::ref(x)); // 传递变量的引用
    t.join();
    return 0;
}
```

---

## 2. 传递指针或引用指向局部变量的问题

```cpp
#include <iostream>
#include <thread>

void foo(int* ptr) {
    std::cout << *ptr << std::endl; // 访问已经被销毁的指针
}

int main() {
    int x = 1;
    std::thread t(foo, &x); // 传递指向局部变量的指针
    t.join();
    return 0;
}
```

### 问题分析

局部变量 `x` 在 `main` 函数作用域结束后被销毁，但 `std::thread` 可能尚未运行或访问了无效的地址，导致未定义行为。

### 解决方案

使用 `new` 在堆上分配内存，或者使用 `std::shared_ptr` 进行管理。

#### 使用 new：

```cpp
#include <iostream>
#include <thread>

void foo(int* ptr) {
    std::cout << *ptr << std::endl;
    delete ptr; // 释放内存
}

int main() {
    int* ptr = new int(1);
    std::thread t(foo, ptr);
    t.join();
    return 0;
}
```

#### 使用 shared_ptr：

```cpp
#include <iostream>
#include <thread>
#include <memory>

void foo(std::shared_ptr<int> ptr) {
    std::cout << *ptr << std::endl;
}

int main() {
    auto ptr = std::make_shared<int>(1);
    std::thread t(foo, ptr);
    t.join();
    return 0;
}
```

---

## 3. 传递指针或引用指向已释放的内存的问题

```cpp
#include <iostream>
#include <thread>

void foo(int& x) {
    std::cout << x << std::endl;
}

int main() {
    int* ptr = new int(1);
    std::thread t(foo, *ptr); // 传递已释放的内存
    delete ptr;
    t.join();
    return 0;
}
```

### 问题分析

在线程 `t` 启动前，`ptr` 已被 `delete`，导致 `foo` 访问了已释放的内存，行为未定义。

### 解决方案

确保在线程执行期间，数据仍然有效。

```cpp
#include <iostream>
#include <thread>
#include <functional>

void foo(int& x) {
    std::cout << x << std::endl;
}

int main() {
    int x = 1;
    std::thread t(foo, std::ref(x)); // 传递变量的引用
    t.join();
    return 0;
}
```

---

## 4. 类成员函数作为入口函数，类对象被提前释放

```cpp
#include <iostream>
#include <thread>

class MyClass {
public:
    void func() {
        std::cout << "Thread " << std::this_thread::get_id() << " started" << std::endl;
        // do some work
        std::cout << "Thread " << std::this_thread::get_id() << " finished" << std::endl;
    }
};

int main() {
    MyClass obj;
    std::thread t(&MyClass::func, &obj);
    return 0;
} // obj 被销毁，可能导致线程崩溃
```

### 问题分析

在 `main` 结束时，`obj` 被销毁，导致 `t` 访问已销毁的对象，可能崩溃。

### 解决方案

使用 `std::shared_ptr` 管理生命周期：

```cpp
#include <iostream>
#include <thread>
#include <memory>

class MyClass {
public:
    void func() {
        std::cout << "Thread " << std::this_thread::get_id() << " started" << std::endl;
        std::cout << "Thread " << std::this_thread::get_id() << " finished" << std::endl;
    }
};

int main() {
    auto obj = std::make_shared<MyClass>();
    std::thread t(&MyClass::func, obj);
    t.join();
    return 0;
}
```

---

## 5. 入口函数为类的私有成员函数

```cpp
#include <iostream>
#include <thread>

class MyClass {
private:
    friend void myThreadFunc(MyClass* obj);
    void privateFunc() {
        std::cout << "Thread " << std::this_thread::get_id() << " privateFunc" << std::endl;
    }
};

void myThreadFunc(MyClass* obj) {
    obj->privateFunc();
}

int main() {
    MyClass obj;
    std::thread thread_1(myThreadFunc, &obj);
    thread_1.join();
    return 0;
}
```

### 问题分析

因为 `privateFunc` 是私有的，不能直接传递给 `std::thread`，解决方案是使用**友元函数**来调用私有成员函数。

